### PR TITLE
Rework networking logic

### PR DESCRIPTION
**Pull request by wrefgtzweve**
Lowers the net.WriteUInt for the entity index and player index.

Players will only take up ent index spots 1 to 255 as can be seen here
https://github.com/Facepunch/garrysmod/blob/master/garrysmod/lua/includes/extensions/net.lua#L76-L87

Only ~8k networked entities can exist so i lowered the ent index to 13, https://github.com/Facepunch/garrysmod/blob/master/garrysmod/lua/includes/extensions/net.lua#L56-L64

This PR brings the per entity transfer amount from 89bits to 46bits which should help to signficiantly reduce network strain.
I think that the net usage can further be reduced here:
```lua
    net.WriteUInt(ent.FPPConstraintReasons and ent.FPPConstraintReasons[ply] or ent.FPPCanTouchWhy[ply], 20) -- reasons
```
But im not entirely sure what the max size of this bitflag can be and how it could be lowered.

This PR also adds a new net queue system, currently for example when a prop is spawned a net message is sent twice see: https://github.com/FPtje/Falcos-Prop-protection/pull/325#issuecomment-1975209906. With this new system it'd only be send once in the next tick, this will prevent sending multiple net messages for multiple changes in a single tick.
This system will automatically spread out large net messages over multiple ticks to prevent buffer overflows/sending too much in a single burst.
It uses a timer that automatically removes itself once the queue is empty, could be swapped out for a hook instead.
The reason i lowered the count per message from 5825 to 1400 is because we're sending 46bits while the max net message size is 65533bits so 65533÷46.

So tldr: this will reduce the amount of networking when for example a prop is spawned from `89` bits to `46` bits per message and make it so only 1 net message is send per prop spawn instead of two. So this would result in 75% less networking.

### CODE REVIEW

**Comment by FPtje** [comment]
Thanks! This probably means that the count limit in `FPP.plySendTouchData` can be increased. Could you update that? See https://github.com/FPtje/Falcos-Prop-protection/issues/323#issuecomment-1902636479

Note: I found the limit by going in a listen server, building a table with just one entity many times, and calling `FPP.plySendTouchData` with that. To get the amount quickly, I used a binary search method on the amount of entities.

**Comment by wrefgtzweve** [comment]
While looking into lowering the limit i noticed that for every prop spawned its actually networked twice, I'll see if i can improve this by doing some kind of 1 tick buffer before sending the data

```
Trace:
	1: Line 32	"Trace"		lua/includes/extensions/debug.lua
	2: Line 280	"netWriteEntData"		addons/falcos-prop-protection/lua/fpp/server/ownability.lua
	3: Line 313	"plySendTouchData"		addons/falcos-prop-protection/lua/fpp/server/ownability.lua
	4: Line 224	"recalculateCanTouch"		addons/falcos-prop-protection/lua/fpp/server/ownability.lua
	5: Line 237	"callback"		addons/falcos-prop-protection/lua/fpp/server/ownability.lua
	6: Line 132	"callback"		addons/falcos-prop-protection/lua/fpp/sh_settings.lua
	7: Line 23	"func"		addons/ulib/lua/ulib/shared/cami_ulib.lua
	8: Line 260	"Call"		addons/ulib/lua/includes/modules/hook.lua
	9: Line 274	"PlayerHasAccess"		addons/falcos-prop-protection/lua/libraries/sh_cami.lua
	10: Line 134	"calculatePlayerPrivilege"		addons/falcos-prop-protection/lua/fpp/sh_settings.lua
	11: Line 237	"recalculateCanTouch"		addons/falcos-prop-protection/lua/fpp/server/ownability.lua
	12: Line 57	"CPPISetOwner"		addons/falcos-prop-protection/lua/fpp/sh_cppi.lua
	13: Line 119	"Finish"		addons/falcos-prop-protection/lua/fpp/server/core.lua
	14: Line 252	"GMODSpawnProp"		gamemodes/sandbox/gamemode/commands.lua
	15: Line 61	"nil"		gamemodes/sandbox/gamemode/commands.lua

Trace:
	1: Line 32	"Trace"		lua/includes/extensions/debug.lua
	2: Line 280	"netWriteEntData"		addons/falcos-prop-protection/lua/fpp/server/ownability.lua
	3: Line 313	"plySendTouchData"		addons/falcos-prop-protection/lua/fpp/server/ownability.lua
	4: Line 388	"onEntitiesCreated"		addons/falcos-prop-protection/lua/fpp/server/ownability.lua
	5: Line 396	"nil"		addons/falcos-prop-protection/lua/fpp/server/ownability.lua
```
image

**Comment by wrefgtzweve** [comment]
Okay this might be a bit of a more complicated PR now, i've updated the PR description.

File: lua/fpp/server/ownability.lua [hunk 1]
```diff
@@ -267,51 +267,58 @@ end
...
+        for ply, entities in pairs(touchabilityDataQueue) do
+            if not IsValid(ply) then touchabilityDataQueue[ply] = nil continue end
+
+            local count = 0
+            net.Start("FPP_TouchabilityData")
+            for ent in pairs(entities) do
+                netWriteEntData(ply, ent)
+                count = count + 1
+                touchabilityDataQueue[ply][ent] = nil
+                if count >= 1400 then break end
```
Comment (FPtje) [inline lua/fpp/server/ownability.lua:295]: I'm not sure I understand this number. With your change, the message is smaller, but the limit is smaller. Where does the number 1400 come from?

File: lua/fpp/server/ownability.lua [hunk 2]
```diff
@@ -267,51 +267,58 @@ end
...
 
-function FPP.plySendTouchData(ply, ents)
-    local count = #ents
-
-    if count == 0 then return end
+local touchabilityDataQueue = {}
+local function startNetWriteQueue()
+    if timer.Exists( "FPP_TouchabilityQueue" ) then return end
+
+    timer.Create( "FPP_TouchabilityQueue", 0, 0, function()
```
Comment (FPtje) [inline lua/fpp/server/ownability.lua:285]: This would add a delay of one tick for any data that gets added to the queue. Previously it would send the data immediately. I'm not sure if that will have a noticeable effect when e.g. the server is slow.

File: lua/fpp/server/ownability.lua [hunk 3]
```diff
@@ -267,51 +267,58 @@ end
 Networking
 ---------------------------------------------------------------------------]]
 util.AddNetworkString("FPP_TouchabilityData")
--- Sends 32 + 32 + 5 + 20 = 89 bits of ownership data per entity
+-- Sends 13 + 8 + 5 + 20 = 46 bits of ownership data per entity
 local function netWriteEntData(ply, ent)
-    -- EntIndex for when it's out of the PVS of the player
-    net.WriteUInt(ent:EntIndex(), 32)
+    -- EntIndex for when it's out of the PVS of the player`
```
Comment (FPtje) [inline lua/fpp/server/ownability.lua:272]: Mistyped character

File: lua/fpp/server/ownability.lua [hunk 4]
```diff
@@ -267,51 +267,58 @@ end
...
     net.WriteUInt(ent.FPPRestrictConstraint and ent.FPPRestrictConstraint[ply] or ent.FPPCanTouch[ply], 5) -- touchability information
     net.WriteUInt(ent.FPPConstraintReasons and ent.FPPConstraintReasons[ply] or ent.FPPCanTouchWhy[ply], 20) -- reasons
 end
 
-function FPP.plySendTouchData(ply, ents)
-    local count = #ents
-
-    if count == 0 then return end
+local touchabilityDataQueue = {}
+local function startNetWriteQueue()
```
Comment (FPtje) [inline lua/fpp/server/ownability.lua:282]: With timers I have a fear of funny race conditions. Think of e.g. this function returning because the timer exists, but right after, the other part of the code decides to remove the timer, even though the queue is not worked away yet.

The tricky thing about such bugs is that they're near impossible to reproduce, but somewhere, among the hundreds of servers running FPP, someone will hit it.

Perhaps a simpler solution can be made without creating and removing timers. 

How about the following idea:
Keep track of a per player count of "amount of entities sent this tick to this player". Also keep track of a per player queue. That queue is a table with number indices, and ent data values.

When sending data to a player, it is done immediately, but for every entity it increases the counter. If the counter hits the limit, you put the rest of remainder of the data in the queue.

Every tick, you reset all counters to 0, and you try to work away the queue from low to high index. If in the tick you hit the limit, you just break and leave the rest for the next tick.

File: lua/fpp/server/ownability.lua [hunk 5]
```diff
@@ -267,51 +267,58 @@ end
...
+    end )
+end
+
+function FPP.plySendTouchData(ply, ents)
+    local count = #ents
+    if count == 0 then return end
+
+    touchabilityDataQueue[ply] = touchabilityDataQueue[ply] or {}
+    for _, ent in ipairs(ents) do
+        touchabilityDataQueue[ply][ent] = true
```
Comment (FPtje) [inline lua/fpp/server/ownability.lua:318]: Is a hashmap for the right data structure for a queue? They're optimized for lookup more than iterating, which is what this code is doing.

Tables are fastest with iterating when they have a non-interrupted sequence of keys starting at 1.

**Review by FPtje — COMMENTED** [review]
Thanks for the improvements! I have a lot of questions and comments about the queue, but I like the idea.

**Comment by FPtje** [comment]
By the way, check out this page on queues: https://www.lua.org/pil/11.4.html

An actual queue implementation could perhaps be nicer if it uses metatables, but the core idea of the structure could work well for this use case.

File: lua/fpp/server/ownability.lua (hunk 1 above)
Comment (wrefgtzweve) [reply lua/fpp/server/ownability.lua:295]: It's in the PR description
"The reason i lowered the count per message from 5825 to 1400 is because we're sending 46bits while the max net message size is 65533bits so 65533÷46."
I feel like gmod internally compresses it in some way but im not entirely sure, i figured 1400 ents/s should be fast enough as the pr makes it half the amount it took before anyways

File: lua/fpp/server/ownability.lua (hunk 2 above)
Comment (wrefgtzweve) [reply lua/fpp/server/ownability.lua:285]: I personally couldn't notice it, the reason the one tick is there is to catch multiple calls that would change the data,  https://github.com/FPtje/Falcos-Prop-protection/pull/325#issuecomment-1975209906 currently FPP sends two net messages for 1 prop spawn because it sets multiple values on it, this 1 tick delay aims to solve that and other addons that possibly change ownership data.

File: lua/fpp/server/ownability.lua (hunk 5 above)
Comment (wrefgtzweve) [reply lua/fpp/server/ownability.lua:318]: The reason i did it this way is because it makes the table structure a lot simpler and easier to parse later on, my main performance concern is calling with the queue is having to check the validity of entities/players as they could've become invalid. With this structure i'd have to only check a player once, i guess i could have a table structure where i do something like {ply=[Player][1],ents={ent1,ent2}} which would be indexed.
I personally haven't seen pairs being a significant performance concern though

File: lua/fpp/server/ownability.lua (hunk 4 above)
Comment (wrefgtzweve) [reply lua/fpp/server/ownability.lua:282]: Im not sure if im getting what you're saying so sorry if i misunderstood. But i dont see how this could lead to a race condition, timers execute in the same sync thread as all the other code, so the timer loop will never be going on while `startNetWriteQueue` is being called and vice versa. As for the immediate sending of data that wouldn't fix the issue described here https://github.com/FPtje/Falcos-Prop-protection/pull/325#issuecomment-1975209906

**Comment by FPtje** [comment]
One request for the caching and sending next frame. Is there a way to measure the cost of storing and retrieving this data from the queue? I'm worried that this administration might be more expensive than just sending an entity twice on creation.

For context, upon creation, the one entity is sent to all players. Doing that twice in one tick is annoying, but not the worst thing. There are some events (e.g. settings changes) that cause all entities to be recalculated and sent to everyone. There, a big overhead is a big problem. Besides, in that situation there likely won't be duplicates, since recalculate-all events are luckily very rare. In that case, this optimisation would only cost, and give no gain.

**Comment by wrefgtzweve** [comment]
I did some testing and it's a bit hard to properly test so if you see any issues in this test code let me know.

```lua
local touchabilityDataQueue = {}
local newTime = 0

local touchabilityDataQueue = {}
local function startNetWriteQueue()
    if timer.Exists( "FPP_TouchabilityQueue" ) then return end

    timer.Create( "FPP_TouchabilityQueue", 0, 0, function()
        local timerTime = SysTime()
        for ply, entities in pairs(touchabilityDataQueue) do
            if not IsValid(ply) then touchabilityDataQueue[ply] = nil continue end

            local count = 0
            --net.Start("FPP_TouchabilityData")
            for ent in pairs(entities) do
                --netWriteEntData(ply, ent)
                count = count + 1
                touchabilityDataQueue[ply][ent] = nil
                if count >= 1400 then break end
            end
            --net.WriteBit(1)
            --net.Send(ply)

            if not next(touchabilityDataQueue[ply]) then
                touchabilityDataQueue[ply] = nil
            end
        end

        if not next(touchabilityDataQueue) then
            timer.Remove( "FPP_TouchabilityQueue" )
            newTime = newTime + SysTime() - timerTime
            print( "New:", newTime )
            print( "-----" )

            return
        end
        newTime = newTime + SysTime() - timerTime
    end )
end

local function sendEntsNew(ply, ents)
    local count = #ents
    if count == 0 then return end

    touchabilityDataQueue[ply] = touchabilityDataQueue[ply] or {}
    for _, ent in ipairs(ents) do
        touchabilityDataQueue[ply][ent] = true
    end

    startNetWriteQueue()
end

local function sendEntsOld(ply, ents)
    local count = #ents

    if count == 0 then return end

    local count_limit = 5825
    if count > count_limit then
        local accumulator = {}
        for i = 1, count do
            table.insert(accumulator, ents[i])
            if i % count_limit == 0 then
                sendEntsOld(ply, accumulator)
                table.Empty(accumulator)
            end
        end
        if #accumulator > 0 then
            sendEntsOld(ply, accumulator)
        end
        return
    end

    --net.Start("FPP_TouchabilityData")
        for i = 1, count do
            --netWriteEntData(ply, ents[i])
            --net.WriteBit(i == count)
        end
    --net.Send(ply)
end

local entities = {}

for i = 1, 8192 do
    table.insert( entities, tostring({}) )
end

print( #entities )


local time1 = SysTime()
sendEntsNew(me, entities)
newTime = newTime + SysTime() - time1

local time2 = SysTime()
sendEntsOld(me, entities)
print( "Old:", SysTime() - time2 )
```

Output:
```
8192
Old:	0.00028520000000753
New:	0.0016020000000481
-----
8192
Old:	0.00028980000001866
New:	0.0016468000001169
-----
8192
Old:	0.00028339999994387
New:	0.001515000000154
-----
8192
Old:	0.00030690000016875
New:	0.0022936999998819
-----
8192
Old:	0.00038549999999304
New:	0.0036048000001756
-----
8192
Old:	0.00028689999999187
New:	0.0018644999995558
-----
8192
Old:	0.00035469999988891
New:	0.0017913000001499
-----
```
So looking at this, for 8192 unique entities (the max amount possible) it's a ~5x to 10x performance increase. This might seem a problem but in reality its not noticable.

A more realistic and fair comparison would be a case where duplicates exist as the new system eliminates it.
```lua
for i = 1, 8192 do
    table.insert( entities, math.random(1,2) == 1 and "x" or "y" )
end
```
So i changed the test data to not be random like here.
Output:
```
8192
Old:	0.00029440000002978
New:	9.7600000117382e-05
-----
8192
Old:	0.00032839999994394
New:	6.800000005569e-05
-----
8192
Old:	0.00030820000006315
New:	0.00010489999999663
-----
8192
Old:	0.00032630000009704
New:	0.00010139999994863
-----
8192
Old:	0.00038440000002993
New:	9.2399999857662e-05
-----
8192
Old:	0.00032759999999143
New:	5.6299999869225e-05
-----
8192
Old:	0.00032249999981104
New:	4.9499999931868e-05
-----
```
So with duplicate data it is significantly faster.
I'd argue that the performance difference in these cases cancels out and shouldn't be a concern.

**Comment by wrefgtzweve** [comment]
I understand that this PR is a big change so I can try this out on some servers for a while, i can for example test this on a ~40-50 player server for a few days to make sure everything works.

**Comment by FPtje** [comment]
That sounds good. It's nice to check a core change like this against a real load

**Comment by FPtje** [comment]
About the double sending of spawned entities, I fixed that in a different way in https://github.com/FPtje/Falcos-Prop-protection/commit/d7f8cda4d41e29456a23b34396488f5094481de5

Essentially the `onEntitiesCreated` would always send the data, even if it didn't change. That's easily fixed, because `FPP.calculateCanTouch` returns whether touchability changed.

Perhaps we can prevent double sending the same way if it happens in other places. I would prefer that over the queue of all information that is sent. That queue is `O(players * entities)` sized, which probably isn't cheap when things are sent for many players and many entities.

What do you think?

**Comment by wrefgtzweve** [comment]
Achieves basically the same as the queue system without making the code significantly more complex. Seems like a better way to go about it yeah.
I do feel like you could apply that change in `plySendTouchData` so it'd apply for all possible changes. Not sure if that'd be bad performance wise though as it'd have to run `calculateCanTouch` each time. So maybe store the last sent state on the entity and compare it against that, although it'd need to store that for each player which would get messy fast too.

But yeah overall, seems better than my solution. The WriteUInt changes and GetTable() are still useful though, i can probably make a separate PR for those to keep things simple, up to you.

**Comment by FPtje** [comment]
Hmm, adding it to `plySendTouchData` would mean that the function no longer does what it says. It also does the calculation. I think part of the strength in FPP is in separating calculation from sending.

I think the other changes in this PR are indeed still useful. You could change this PR, or close this PR in favor of a new one. Either is fine with me 👍

**Comment by wrefgtzweve** [comment]
Made a new PR with the UInt improvements https://github.com/FPtje/Falcos-Prop-protection/pull/330
